### ISSUE TITLE
Missing Endpoints after ADS stream re-connection

### ISSUE THREAD

**Issue by freeloyerkick** [envoyproxy/envoy#13072]
*Title*:  Missing Endpoints after ADS stream re-connection 

*Description*:

We are using the more recent Envoy version (1.14.1) with most recent envoyproxy/go-control-plane as its ADS control plane. Recently we observed an production incident that, upon deployment of the control plane, some percentage of Envoys has empty endpoints for all known clusters (verified via Envoy /clusters endpoint, all clusters are there, but no Endpoints), causing 503 for all incoming requests. Restarting of Envoy fixes the issue.

Looking at the Envoy log around that time:

```
[2020-09-12 02:19:33.038][15744][warning][config] [bazel-out/k8-opt/bin/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:92] StreamAggregatedResources gRPC config stream closed: 2, grpc recei
ve from proxy client failed: rpc error: code = Unavailable desc = transport is closing
[2020-09-12 02:19:38.265][15744][warning][config] [bazel-out/k8-opt/bin/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:92] StreamAggregatedResources gRPC config stream closed: 2, failed to
get a client from the ClientFactory: failed to obtain client connection: failed to dail xds peer 10.90.171.157:31003: context deadline exceeded
[2020-09-12 02:19:47.152][15744][info][upstream] [source/common/upstream/cds_api_impl.cc:77] cds: add 1755 cluster(s), remove 1 cluster(s)
[2020-09-12 02:19:47.153][15744][info][upstream] [source/common/upstream/cds_api_impl.cc:93] cds: add/update cluster 'foo'
[2020-09-12 02:19:47.153][15744][info][upstream] [source/common/upstream/cds_api_impl.cc:93] cds: add/update cluster 'bar'
```
apparently the ADS stream was re-connected (due to ADS server restart), and then immediately received CDS update with all the clusters. 

So I have two questions here that would help us debug further:

1. When the new clusters (CDS) are received by Envoy, will it override all existing cluster & clusterLoadAssignment resources, and hence zero out all endpoints before an EDS update were received?? That'll be really bad, which could be the cause of the issue, can someone help explain?

2. Separately, to protect against existing endpoints being zero-ed out completely and cause outages, is there a way in Envoy to guard against that, i.e. keep the endpoint from the last EDS update, if the new EDS update has zero endpoints?

**Comment by htuch** [comment]
I think this might be overlap with https://github.com/envoyproxy/envoy/issues/13009.

**Comment by fxposter** [comment]
In addition to that issue https://github.com/envoyproxy/go-control-plane/pull/353 can be the cause for it

**Comment by freeloyerkick** [comment]
Thanks for pointing the related issues, I think those may be related, we did see CPU throttling of the control plane (using envoyproxy/go-control-plane) upon the incident, which could be the trigger of missing EDS update. And the fact that go-control-plane doesn't handle NACK makes the issue even worse. (we also observed that after a new xds snapshhot was taken in the control plane which triggered a new EDS update, issue got resolved)

separately, this is a severe issue, and I am still looking for mitigation strategies to prevent this from happening --- empty endpoints causing 503. 

e.g. as part of CDS update, is there a way to configure a (static) fallback endpoint for a cluster when there is no healthy endpoints (or no endpoints due to missing EDS update).
